Issue/13066 ppn categories

zwarm opened the pull request:
Part of #13066 

This PR adds support for
1. Selecting categories for a post
2. Adding a new category

This PR is labeled WIP/do not merge. I am looking for a review before moving forward with the open issues (see list below).

The meat of the PR can be found in:
`PrepublishingAddCategoryFragment` and `PrepublishingAddCategoryViewModel`
- This pair is responsible for the adding of a new category 

`PrepublishingCategoriesFragment and PrepublishingCategoriesViewModel`
- This pair is responsible for selecting/unselecting categories and for kicking off the Add Category action.

`PrepublishingBottomSheetFragment` (preexisting) is the traffic cop and handles adding/removing fragment transactions to the bottom sheet container. This was updated to handle the AddCategory and Categories fragments. Each of the individual actions in the bottom sheet are standalone fragments and the behavior is identical for all.

I was on the fence with regards to suspendable functions.
- They are used for update/add/fetch server calls in the `UpdatePostCategoriesUseCase`, `AddCategoryUseCase`, and `GetCategoriesUseCase`. I followed existing architecture for calls to TaxonomyStore and EditPostRepository. I am open to suggestions.

It’s a rather larger PR, but I broke it up by commits.
f8a263b => layout files, animation files, strings, `AppComponent`/`ViewModelModule`, plugging into existing architecture
740ce31 => Changes to `PrepublishingBottomSheetFragmentPrepublishingHomeViewModel`, `PrepublishingViewModel`
7c5012d => Add category related
3d28bb4 => Select category related

Home|Select|Add
---|---|---
<img width="480" src="https://user-images.githubusercontent.com/506707/96194465-6995f880-0f18-11eb-8959-d21eea29fd6f.png">|<img width="480" src="https://user-images.githubusercontent.com/506707/96194480-73b7f700-0f18-11eb-9202-25a09d2ba0a2.png">|<img width="480" src="https://user-images.githubusercontent.com/506707/96194499-7c103200-0f18-11eb-8825-2d9cca288bdf.png">

**This PR has open issues** and they are almost all related to BottomSheet behavior (scrolling the biggest one)
1. Showing a snackbar in a bottom sheet (using toast for now)
~~2. Recyclerview: scrolling within a bottom sheet (this also affects PTR)~~  ****Update**: Removed the PTR in commit cd27ec0 and scrolling works just fine. 
3. There is an existing [issue](https://github.com/wordpress-mobile/WordPress-Android/issues/12166) for the janky transitions between the bottom sheet fragments. I started on some new transition files and things were looking positive, but ultimately decided to push this issue off until all the bottom sheet scrolling issues are resolved.
4. There is a separate [issue](https://github.com/wordpress-mobile/WordPress-Android/issues/13149) for writing unit test 
I am open to any suggestions for solving the scrolling/recyclerview issues in a bottom sheet. 😉 

To test:
Prep
- Launch the app
- Select Blog posts
- Select Edit on a post
- From the edit post view, tap the UPDATE action to launch the bottom sheet

Categories are shown on the Categories row
- If the post has categories assigned, you will see them listed on the home view (figure 1 above). If not, you will see Not Set

Select Categories is shown when user taps on Categories
- Tap the categories row to launch the Select categories view (fig 2)

Add category is shown and user can save a new category
- Tap the Add New Category button on the bottom of the select categories view (fig 3)
- Enter a new category and tap the <- Back Arrow

Newly Added category is shown in the categories list
- Take note of the toast which will tell you if the action succeeded or failed. 
- Back on the select category view, the category you have added is visible and checked

Categories can be checked/unchecked 
- Select or unselect some categories

Categories are saved
- Tap the <- Back arrow to save your options
- The categories row on the home screen should reflect the selections you made

PR submission checklist:

- [X] I have considered adding unit tests where possible.
- [X] I have considered adding accessibility improvements for my changes.
- [X] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

zwarm committed:
Add close button to toolbar

zwarm committed:
WIP

zwarm committed:
Add prepublish categories

zwarm committed:
Add categories support

zwarm committed:
Support for Add category

zwarm committed:
Add select categories

malinajirka commented on WordPress/src/main/java/org/wordpress/android/ui/posts/GetCategoriesUseCase.kt:
```diff
@@ -0,0 +1,62 @@
+        editPostRepository: EditPostRepository,
+        siteModel: SiteModel
+    ): String? {
+        val post = editPostRepository.getPost() ?: return null
```
We might want to log an exception - since I believe the null state is not expected, right?

zwarm replied:
Exception is now logged in c42360b

malinajirka commented on WordPress/src/main/java/org/wordpress/android/ui/posts/GetCategoriesUseCase.kt:
```diff
@@ -0,0 +1,62 @@
+        val rootCategory = CategoryNode.createCategoryTreeFromList(
+                getCategoriesForSite(siteModel)
+        )
+        return CategoryNode.getSortedListOfCategoriesFromRoot(rootCategory) ?: arrayListOf()
```
We might want to create a wrapper around CategoryNode, since the static access is not testable. Wdyt?

zwarm replied:
Wrapper created in c389ca3

malinajirka commented on WordPress/src/main/java/org/wordpress/android/ui/posts/GetCategoriesUseCase.kt:
```diff
@@ -0,0 +1,62 @@
+    }
+
+    suspend fun fetchSiteCategories(siteModel: SiteModel) {
+        withContext(ioDispatcher) {
```
The dispatch action is asynchronous. We might want to consider removing the "suspend and withContext". The context switching is most probably more expensive than the action itself. Wdyt?

zwarm replied:
Makes sense in 3d8941d.

malinajirka commented on WordPress/src/main/java/org/wordpress/android/ui/posts/GetCategoriesUseCase.kt:
```diff
@@ -0,0 +1,62 @@
+    }
+
+    private fun formatCategories(categoryList: List<TermModel>): String? {
+        if (categoryList.isEmpty()) return null
```
I don't have the full context, but could we return an empty string here and change the return type to non-nullable string?

zwarm replied:
Returning empty string in 3055751

malinajirka commented on WordPress/src/main/java/org/wordpress/android/ui/posts/PrepublishingAddCategoryFragment.kt:
```diff
@@ -0,0 +1,165 @@
+    override fun onResume() {
+        val needsRequestLayout = requireArguments().getBoolean(NEEDS_REQUEST_LAYOUT)
+        if (needsRequestLayout) {
+            requireActivity().window.decorView.requestLayout()
```
Could you please elaborate why we need to manually request layout? We might want to consider adding a comment here. Wdyt?

zwarm replied:
Will add comment

zwarm replied:
This was introduced when Stories were added. Specific to PPN it's so the RecyclerView is re-calculated and visible when coming from stories. The functionality was added to AddCategory and Category fragments to maintain usability for coming from stories.

zwarm replied:
Comment added in 9216cca

malinajirka commented on WordPress/src/main/java/org/wordpress/android/ui/posts/PrepublishingAddCategoryFragment.kt:
```diff
@@ -0,0 +1,165 @@
+        super.onResume()
+    }
+
+    companion object {
```
NitPick: We usually keep the companion object at the bottom of the file.

zwarm replied:
in 9216cca

malinajirka commented on WordPress/src/main/java/org/wordpress/android/ui/posts/PrepublishingAddCategoryFragment.kt:
```diff
@@ -0,0 +1,165 @@
+
+    private fun initBackButton() {
+        back_button.setOnClickListener {
+            viewModel.addCategory(
```
IMO the view layer shouldn't be responsible for such logic. It should just propagate user and system events to the VM. 

Ideally we'd just invoke `viewModel.onBackButtonClicked()` and the VM would do whatever is necessary. The VM should already know the state of editable fields. If it doesn't I guess we could propagate it as a parameter in "onBackButtonClicked" but it'd still feel a bit weird.

Wdyt?

zwarm replied:
Moved to #13066  ticket to track as a separate item outside of this PR

malinajirka commented on WordPress/src/main/java/org/wordpress/android/ui/posts/PrepublishingAddCategoryFragment.kt:
```diff
@@ -0,0 +1,165 @@
+
+    private fun initCloseButton() {
+        close_button.setOnClickListener {
+            viewModel.onBackButtonClicked()
```
I'd consider renaming "onBackButtonClicked" to something else - since we use it for both "back" and "close". Or we can introduce a new method for "onCloseButtonClicked". Wdyt?

zwarm replied:
The close button is being removed and a true Submit button will be added. See issue ticket for details #13066

malinajirka commented on WordPress/src/main/java/org/wordpress/android/ui/posts/PrepublishingAddCategoryFragment.kt:
```diff
@@ -0,0 +1,165 @@
+import kotlinx.android.synthetic.main.add_category.*
+import kotlinx.android.synthetic.main.prepublishing_toolbar.*
+import org.wordpress.android.R
+import org.wordpress.android.R.layout
```
Lint is having hard time with all "R.xyz" imports - it doesn't find them and complains after, that a resource is unused. I'd consider removing it. Wdyt?

zwarm replied:
Fixed in 646567b

malinajirka commented on WordPress/src/main/java/org/wordpress/android/ui/posts/PrepublishingAddCategoryFragment.kt:
```diff
@@ -0,0 +1,165 @@
+        val message = uiHelpers.getTextOfUiString(requireContext(), this.message).toString()
+        ToastUtils.showToast(
+                requireContext(), message,
+                LONG
```
Is there any particular reason why we want to go with "LONG" instead of "SHORT" here? AFAIK we use SHORT for most usecases in the app. Wdyt?

zwarm replied:
I switched to Long because the message doesn't show in the view in which the action was submitted. I wanted the user to have more time to notice if there was an issue.

malinajirka commented on WordPress/src/main/java/org/wordpress/android/ui/posts/PrepublishingAddCategoryViewModel.kt:
```diff
@@ -0,0 +1,124 @@
+import kotlinx.coroutines.CoroutineDispatcher
+import kotlinx.coroutines.Job
+import kotlinx.coroutines.launch
+import org.wordpress.android.R.string
```
Same as one of the comments above - I'd consider removing the ".string" part from the import.

zwarm replied:
Fixed in 646567b

malinajirka commented on WordPress/src/main/java/org/wordpress/android/ui/posts/PrepublishingCategoriesViewModel.kt:
```diff
@@ -0,0 +1,223 @@
+    private val _navigateToAddCategoryScreen = MutableLiveData<Event<Unit>>()
+    val navigateToAddCategoryScreen: LiveData<Event<Unit>> = _navigateToAddCategoryScreen
+
+    private val _toolbarTitleUiState = MutableLiveData<UiString>()
```
Out of curiosity - is there any reason why you prefer having multiple uiStates (toolbarUiState, refreshingUiState, uiState) over having a single uistate? Both approaches (single vs multiple) are valid and used in our codebase. 

I always prefer having a single UIState as it's extremely simple to make sure that an invalid state even cannot be created. It's also way easier to write tests for it since there are clearly defined combinations of possible states.

zwarm replied:
I lifted the logic for the toolbar/navigate from the existing PPN fragments/viewmodels and in the interest of time, used them as they were written. I'll jots this down as a nice-to-change list when I work on some overall architecture changes. Not part of this PR

zwarm replied:
Moved to #13066  ticket to track as a separate item outside of this PR

malinajirka commented on WordPress/src/main/java/org/wordpress/android/ui/posts/PrepublishingCategoriesViewModel.kt:
```diff
@@ -0,0 +1,223 @@
+    val refreshingUiState: LiveData<SwipeToRefreshUiState> = _refreshingUiState
+
+    fun start(editPostRepository: EditPostRepository, siteModel: SiteModel) {
+        this.editPostRepository = editPostRepository
```
Won't this throw an exception? Both editPostRepository and siteModel are lateinit vars, and if I recall correctly, those can't be re-assigned. Wdyt?

zwarm replied:
Noted in #13066 under arch changes.

zwarm replied:
Actually @malinajirka I think those are fine - hence the use of the `this` keyword. They are being initialized not reassigned. I've seen this pattern in [Interests](https://github.com/wordpress-mobile/WordPress-Android/blob/2476981600502ddb9a8c73207e937fc1edb48707/WordPress/src/main/java/org/wordpress/android/ui/reader/discover/interests/ReaderInterestsViewModel.kt#L53), [Stories](https://github.com/wordpress-mobile/WordPress-Android/blob/2476981600502ddb9a8c73207e937fc1edb48707/WordPress/src/main/java/org/wordpress/android/ui/stories/StoryComposerViewModel.kt#L64), [PostList](https://github.com/wordpress-mobile/WordPress-Android/blob/2476981600502ddb9a8c73207e937fc1edb48707/WordPress/src/main/java/org/wordpress/android/ui/posts/PostListMainViewModel.kt#L227), [Tags](https://github.com/wordpress-mobile/WordPress-Android/blob/2476981600502ddb9a8c73207e937fc1edb48707/WordPress/src/main/java/org/wordpress/android/ui/posts/PrepublishingTagsViewModel.kt#L39)

I mean, we could all be wrong 🤣  The other option is setting them var xxx = null? Thoughts?

malinajirka replied:
Isn't the `this` keyword used because of the name collision - both the parameter and the field are called `editPostRepository`?

> They are being initialized not reassigned.

But start() can be called multiple times, right? So the first call is "initialization" but any subsequent calls are "reassignments".

Either way, I've created a unit test and it seems "lateinit vars" can be re-assigned. Which kind of make sense, otherwise, they'd be called "lateinit val". So we are good - even though it's probably not a great practice since we are opening doors for race conditions. Having said that it's completely fine to use it here - we already use it at other places as you mentioned.

zwarm replied:
Ah, yes, I forgot about `start` being called over and over again. It does seem like a poor pattern to rely on, I'll drop it on my list for hack week when I consolidate editPostSettings in. Thanks

malinajirka commented on WordPress/src/main/java/org/wordpress/android/ui/posts/PrepublishingCategoriesViewModel.kt:
```diff
@@ -0,0 +1,223 @@
+    }
+
+    private fun setToolbarTitleUiState() {
+        _toolbarTitleUiState.postValue(UiStringRes(string.prepublishing_nudges_toolbar_title_categories))
```
Just out of curiosity - any reason why you use "postValue" instead of "value = " here?

zwarm replied:
Same comment as above: I lifted the logic for the toolbar/navigate from the existing PPN fragments/viewmodels and in the interest of time, used them as they were written.  Noted in the issue #13066.

malinajirka commented on WordPress/src/main/java/org/wordpress/android/ui/posts/PrepublishingCategoriesViewModel.kt:
```diff
@@ -0,0 +1,223 @@
+    fun updateCategories() {
+        if (!hasChanges()) return
+
+        if (!networkUtilsWrapper.isNetworkAvailable()) {
```
AFAIU we call this method from "backPressed" and that's the only path how to save the changes. When the user presses "back" in offline, won't we just throw away their changes?

zwarm replied:
I'm not sure what the desired behavior is and interested in what your thoughts are? I'm going to check how the setting work, maybe that will provide a hint. The PPN tags updates each keypress (with a delay), so that doesn't happen on back button. I didn't want to do that here because we might overload categories. wdyt?

malinajirka replied:
It works the same way in the settings, so I think we are good. Having said that, I think it's not a great UX - we probably should display a dialog and let the user confirm that they are ok loosing their changes. Either way, it's out of scope.

zwarm replied:
> I think it's not a great UX

I'm with you on this ... no notifications success/failure ... I get it, but we are all over the place in the app and we probably won't come to a standard. 🤷

malinajirka commented on WordPress/src/main/java/org/wordpress/android/ui/posts/PrepublishingCategoriesViewModel.kt:
```diff
@@ -0,0 +1,223 @@
+
+    override fun onCleared() {
+        super.onCleared()
+        updateCategoriesJob?.cancel()
```
Just to double check: Do we really want to cancel the saving operation when the system destroys the VM?

zwarm replied:
I followed the strategy in the other PPN actions figuring I should match what is already being done. If we change here, should we change the other PPN actions to match? @malinajirka

malinajirka replied:
If it's consistent with the rest I think it's okay to keep it - we'd ideally update all the actions, but it's above the scope. Thanks for the clarification!

zwarm replied:
FYI: I pulled it out in my refactors :)

malinajirka replied:
Awesome, thanks!! ❤️

malinajirka commented on WordPress/src/main/java/org/wordpress/android/ui/posts/PrepublishingCategoriesViewModel.kt:
```diff
@@ -0,0 +1,223 @@
+        updateCategoriesJob = launch(bgDispatcher) {
+            updatePostCategoriesUseCase.updateCategories(categoryList.toList(), editPostRepository)
+        }
+        trackCategoriesChangedEvent()
```
This line will be executed no matter if the updateCategoriesJob started/finished. Is this intentional? Don't we want to track the event after the categories are saved?

zwarm replied:
Ugh, I suppose I can move it to the home fragment since we don't wait around for success or failure. The same would be true for add category - the capture would have to occur in the select logic. Bleh. Noted in the issue #13066.

malinajirka commented on WordPress/src/main/java/org/wordpress/android/ui/posts/PrepublishingCategoriesViewModel.kt:
```diff
@@ -0,0 +1,223 @@
+    fun addSelectedCategory(categoryId: Long) {
+        uiState.value.let { state ->
+            if (state is ContentUiState) {
+                state.selectedCategoryIds.add(categoryId)
```
Have you considered using an immutable collection? Using a mutable uiState intentionally feels a bit risky. Wdyt?

zwarm replied:
Noted in the issue #13066 under arch changes.

malinajirka replied:
Could you please double check this is in the list? I can't find it there. I think we use it in some others VMs too, but it might bite us.

malinajirka commented on WordPress/src/main/java/org/wordpress/android/ui/posts/PrepublishingCategoriesViewModel.kt:
```diff
@@ -0,0 +1,223 @@
+    fun removeSelectedCategory(categoryId: Long) {
+        uiState.value.let { state ->
+            if (state is ContentUiState) {
+                state.selectedCategoryIds.add(categoryId)
```
Same as above.

zwarm replied:
Noted in the issue #13066 under arch changes.

malinajirka commented on WordPress/src/main/java/org/wordpress/android/ui/posts/PrepublishingAddCategoryViewModel.kt:
```diff
@@ -0,0 +1,124 @@
+        closeKeyboard: Boolean = false
+    ) {
+        this.closeKeyboard = closeKeyboard
+        this.siteModel = siteModel
```
Similar as above - we are re-assigning a lateinit var.

zwarm replied:
Noted in the issue #13066 under arch changes.

malinajirka commented on WordPress/src/main/java/org/wordpress/android/ui/posts/PrepublishingAddCategoryViewModel.kt:
```diff
@@ -0,0 +1,124 @@
+        updateUiState(newUiState)
+    }
+
+    fun addCategory(categoryText: String, parentCategory: CategoryNode?) {
```
Similar as in the previous VM
- AFAIU we'll just throw away user changes in offline
- do we really want to cancel the saving operation in `onCleared` ?

zwarm replied:
Same as above, I followed the strategy in the other PPN actions figuring I should match what is already being done. If we change here, should we change the other PPN actions to match?

malinajirka commented on WordPress/src/main/java/org/wordpress/android/ui/posts/PrepublishingAddCategoryViewModel.kt:
```diff
@@ -0,0 +1,124 @@
+        analyticsTrackerWrapper.trackPrepublishingNudges(Stat.EDITOR_POST_CATEGORIES_ADDED)
+    }
+
+    private fun updateUiState(uiState: UiState) {
```
We might want to consider inlining this method - it's used just from a single place and it's body is pretty self-explanatory. Wdyt?

zwarm replied:
Noted in the issue #13066 under arch changes, but it might be moot if the tracking location changes to the select categories view.

malinajirka commented on WordPress/src/main/java/org/wordpress/android/ui/posts/PrepublishingBottomSheetFragment.kt:
```diff
@@ -190,12 +210,10 @@ class PrepublishingBottomSheetFragment : WPBottomSheetDialogFragment(),
     private fun fadeInFragment(fragment: Fragment, tag: String) {
         childFragmentManager.let { fragmentManager ->
             val fragmentTransaction = fragmentManager.beginTransaction()
-            fragmentManager.findFragmentById(R.id.prepublishing_content_fragment)?.run {
```
Could you please elaborate why we are removing this line?

zwarm replied:
This was a fantastic catch. I removed it by accident because I was rushing to swap out the transitions at the last moment.

zwarm replied:
In 646567b

malinajirka commented on WordPress/src/main/java/org/wordpress/android/ui/posts/PrepublishingCategoriesAdapter.kt:
```diff
@@ -0,0 +1,116 @@
+        return categoryNodeList.size
+    }
+
+    private fun isSameCategoryList(categoryList: List<CategoryNode>): Boolean {
```
Have you considered implementing DiffUtils instead of a custom setter?

zwarm replied:
Yes, it's on my short list, but didn't transition it over before submitting 1st review.

zwarm replied:
Noted in the issue #13066 under arch changes.

malinajirka commented on WordPress/src/main/java/org/wordpress/android/ui/posts/PrepublishingCategoriesAdapter.kt:
```diff
@@ -0,0 +1,116 @@
+        return false
+    }
+
+    class CategoriesViewHolder(
```
We might want to consider extracting this class into a different file. I think it'd be consistent with the other VH. Wdyt?

zwarm replied:
Noted in the issue #13066 under arch change

malinajirka commented on WordPress/src/main/java/org/wordpress/android/ui/posts/PrepublishingCategoriesAdapter.kt:
```diff
@@ -0,0 +1,116 @@
+            )
+
+            prepublishing_category_check.isChecked = false
+            if (selectedCategoryIds.contains(row.categoryId)) {
```
Have you considered keeping the "check/not-checked" state within the UiState? Currently the view layer is responsible for checking if `selectedCategoryIds` contains the item which is about to be rendered. It feels a bit weird. Ideally the VM would create the UiStates and the only job of the VH would be to render the UI state. The UIState could even contain `vertical padding = R.dimen.margin_large`, `horizontalPadding = R.dimen.margin_extra_large` and `start padding multiplier = row.level`. Wdyt?

zwarm replied:
🤔  this is a good suggestion. I like it.

malinajirka commented on WordPress/src/main/java/org/wordpress/android/ui/posts/PrepublishingCategoriesFragment.kt:
```diff
@@ -0,0 +1,239 @@
+    private fun initRecyclerView() {
+        recycler_view.layoutManager = LinearLayoutManager(context, RecyclerView.VERTICAL, false)
+        recycler_view.adapter = PrepublishingCategoriesAdapter(
+                onCheckChangeListener = { id, checked ->
```
Same as the above comments - the view layers shouldn't be responsible for such logic. Have you considered storing the listener in the uiState of the items?

zwarm replied:
Noted in the issue #13066 under arch change

malinajirka replied:
Which of the items in the issue is referring to this comment?

malinajirka commented on WordPress/src/main/java/org/wordpress/android/ui/posts/PrepublishingCategoriesFragment.kt:
```diff
@@ -0,0 +1,239 @@
+
+    @SuppressWarnings("unused")
+    @Subscribe(threadMode = MAIN)
+    fun onTaxonomyChanged(event: OnTaxonomyChanged) {
```
Have you considered observing these changes in the VM layer not the view layer?

zwarm replied:
Noted in the issue #13066 under arch change

malinajirka commented on WordPress/src/main/java/org/wordpress/android/ui/posts/PrepublishingCategoriesFragment.kt:
```diff
@@ -0,0 +1,239 @@
+
+    private fun initBackButton() {
+        back_button.setOnClickListener {
+            viewModel.updateCategories()
```
Same as for the other fragment -  we might want to consider handling this within the VM. The view layer shouldn't tell the VM what to do (eg. it shouldn't tell the VM to updateCategories). It's not testable. Wdyt?

zwarm replied:
Noted in the issue #13066 under arch change

malinajirka commented on WordPress/src/main/java/org/wordpress/android/ui/posts/PrepublishingCategoriesFragment.kt:
```diff
@@ -0,0 +1,239 @@
+    private fun initSwipeToRefreshHelper() {
+        swipeToRefreshHelper = WPSwipeToRefreshHelper.buildSwipeToRefreshHelper(ptr_layout,
+                SwipeToRefreshHelper.RefreshListener {
+                    if (!NetworkUtils.checkConnection(requireContext())) {
```
Have you considered moving this logic into the VM?

zwarm replied:
Going to remove PTR all together because it doesn't belong in a bottom sheet. 👍

zwarm replied:
Separate task in the issue #13066.

zwarm replied:
in commit cd27ec0
After I removed the SwipeRefreshLayout, miraculously the scrolling of the recyclerview works like a charm. 🥳

malinajirka commented on WordPress/src/main/res/anim/prepublishing_fragment_fade_in.xml:
```diff
@@ -1,10 +1,14 @@
-        android:startOffset="200"
-        android:duration="200" />
+        android:toAlpha="1" />
+<!--<alpha-->
```
Is this change intentional - if it is, we might want to remove the commented out code. Wdyt?

zwarm replied:
In 89a9945

malinajirka commented on WordPress/src/main/res/anim/prepublishing_fragment_fade_out.xml:
```diff
@@ -1,10 +1,14 @@
-        android:startOffset="200"
-        android:duration="200" />
+        android:toAlpha="0" />
+<!--<alpha-->
```
Same as above.

zwarm replied:
In 89a9945

malinajirka reviewed (commented):
Thanks @zwarm! 

I've done my best reviewing the PR, but it's huge. Please try to keep PRs focused. This PR could have been easily split into 4-6 single focused PRs and the review would be way easier.

It seems you've decided not to re-use any part of the existing "add category" screen from the Post settings. It was probably a good decision. However, I'm wondering -  have we considered using this new code even for that screen? 

> I am looking for a review before moving forward with the open issues (see list below).

I'd suggest fixing the issues in separate PRs. Wdyt?

> Recyclerview: scrolling within a bottom sheet (this also affects PTR)
>There is an existing issue for the janky transitions between the bottom sheet fragments. I started on some new transition files and things were looking positive, but ultimately decided to push this issue off until all the bottom sheet scrolling issues are resolved.

This is the cost of mis-using bottom sheets for something they weren't designed for :D. I'd suggest not providing PTR in bottom sheets. Wdyt?




Bugs:
1. The list items are not clickable - I need to click right on the check box.
2. I can't de-select categories and save the changes. When I de-select them and click on back, they are still there.
<img src="https://user-images.githubusercontent.com/2261188/96248111-5bdd8300-0fab-11eb-8730-23b8e91c21c6.gif" width="250px" />

3. When I select/de-select a category and rotate my device, the change is lost.
<img src="https://user-images.githubusercontent.com/2261188/96248179-73b50700-0fab-11eb-9950-c32b389b29af.gif" width="250px" />
4. When I set parent category and rotate my device, the change is lost.

<img src="https://user-images.githubusercontent.com/2261188/96248289-9a733d80-0fab-11eb-913c-e6f8ee39e84b.gif" width="250px" />

5. When I enter a new category and go back -> it redirects me to the list of categories, but the new category is not there and after +-1s it redirects me back to "add new category screen" when I click back again, the category appears in the list.

<img src="https://user-images.githubusercontent.com/2261188/96248312-a828c300-0fab-11eb-90da-fb24891ace36.gif" width="250px" />


6. (@mbshakti) When I clicked on the close icon on the "add category screen" I expected the whole bottom sheet would get closed. If we want to treat it as "discard" I'd consider a different icon/text. Wdyt?

(Emulator Pixel 3a, API 29)

zwarm commented:
> I've done my best reviewing the PR, but it's huge. Please try to keep PRs focused. This PR could have been easily split into 4-6 single focused PRs and the review would be way easier.

Yeah, I don't know what I was thinking. I committed on the branch to switch context earlier in the week and from there it went down hill fast. I'll switch this merge from develop to a feature branch. It will be easier to track.

> It seems you've decided not to re-use any part of the existing "add category" screen from the Post settings. It was probably a good decision. However, I'm wondering - have we considered using this new code even for that screen?

Yes and yes. I did lift layout files from old code, but didn't want to lock us into listview. My next hack week project is going to be refactoring the existing Post Settings categories and bring it in line with this. 
 
> I'd suggest fixing the issues in separate PRs. Wdyt?
Yes. I'll change the branch name to merge this PR into and go from there.

> This is the cost of mis-using bottom sheets for something they weren't designed for :D. I'd suggest not providing PTR in bottom sheets. Wdyt?
 I agree with you 100%. Even recycler views are crappy in bottom sheets. And since each fragment is not it's own bottom sheet, I'll need to add hooks back to the main bottom sheet fragment for any "on-the-fly" behavior changes. I have some thoughts about making the 'Add New Category' a static bottom component. I'll spend some time on that today. Goodbye PTR!

> Bugs:
> 1. The list items are not clickable - I need to click right on the check box.

I did not make them clickable, but I will.

> 2. I can't de-select categories and save the changes. When I de-select them and click on back, they are still there.
> 3. When I select/de-select a category and rotate my device, the change is lost.

Thanks, I thought I got these all

> 4. When I set parent category and rotate my device, the change is lost.

Will take another look at this one

> 1. When I enter a new category and go back -> it redirects me to the list of categories, but the new category is not there and after +-1s it redirects me back to "add new category screen" when I click back again, the category appears in the list.

That one is odd, I see the new category fairly quickly.

@malinajirka  Thanks for the set of eyes

zwarm committed:
Remove comments

zwarm committed:
Remove suspend function

zwarm committed:
Fix lint and test

zwarm committed:
Add comment.Move companion

zwarm committed:
Log null post

zwarm committed:
Replace null with empty string

zwarm committed:
Add CategoryNodeWrapper

zwarm committed:
Remove PTR functionality

zwarm committed:
Make row selectable and remove when unselected

zwarm commented:
> 1. The list items are not clickable - I need to click right on the check box.

I made the rows clickable in 09250ff

> 2. I can't de-select categories and save the changes. When I de-select them and click on back, they are still there.

Fixed in 09250ff

> 3. When I select/de-select a category and rotate my device, the change is lost.

Fixed in 09250ff

> 4. When I set parent category and rotate my device, the change is lost.

Fixed in 701759b

> 5. When I enter a new category and go back -> it redirects me to the list of categories, but the new category is not there and after +-1s it redirects me back to "add new category screen" when I click back again, the category appears in the list.

Fixed in 040f65b

zwarm committed:
WIP because AS is borked

zwarm committed:
Handle spinner on rotation

zwarm committed:
Fix add category reopen

zwarm committed:
Fix lint

malinajirka reviewed (approved):
Thanks @zwarm! Thanks for the updates, great job ;)! 

I've found just one issue and I don't think it's vital.
1. Select some categories
2. Click on "Add new category"
3. Either press back right way or type something and press back
4. Notice the categories selected in step `1` are not selected.

I think this will get fixed when we include "check/unchecked" state in the uiState of each item. Wdyt?

Feel free to merge this PR. I left some comments, but they are mostly clarification questions, not blockers.

zwarm commented:
> I've found just one issue and I don't think it's vital.
> I think this will get fixed when we include "check/unchecked" state in the uiState of each item. Wdyt?

Yep, cleaned up in the refactor code. I'm going to merge and get to a stabler place for further reviews. Thanks so much @malinajirka for the good feedback.

mbshakti commented:
From a design POV, looks good to me. Thanks @zwarm !
